src/main/scala/is/hail/expr/types/TTuple.scala

```diff
@@ -0,0 +1,63 @@
+package is.hail.expr.types
+
+import is.hail.annotations.ExtendedOrdering
+import is.hail.utils._
+
+import scala.collection.JavaConverters._
+
+object TTuple {
+  private val requiredEmpty = TTuple(Array.empty[Type], true)
+  private val optionalEmpty = TTuple(Array.empty[Type], false)
+
+  def empty(required: Boolean = false): TTuple = if (required) requiredEmpty else optionalEmpty
+
+  def apply(required: Boolean, args: Type*): TTuple = TTuple(args.toArray, required)
+
+  def apply(args: Type*): TTuple = apply(false, args: _*)
+
+  def apply(types: java.util.ArrayList[Type], required: Boolean): TTuple = {
+    val t = TTuple(types.asScala.toArray)
+    t.setRequired(required).asInstanceOf[TTuple]
+  }
+}
+
+final case class TTuple(types: IndexedSeq[Type], override val required: Boolean = false) extends TStructBase {
+
+  val ordering: ExtendedOrdering =
+    ExtendedOrdering.rowOrdering(types.map(_.ordering).toArray)
+
+  def ++(that: TTuple): TTuple = TTuple(types ++ that.types, required = false)
+
+  override def canCompare(other: Type): Boolean = other match {
+    case t: TTuple => size == t.size && types.zip(t.types).forall { case (t1, t2) => t1.canCompare(t2) }
+    case _ => false
+  }
+
+  override def unify(concrete: Type): Boolean = concrete match {
+    case TTuple(ctypes, _) =>
+      size == ctypes.length &&
+        (types, ctypes).zipped.forall { case (t, ct) =>
+          t.unify(ct)
+        }
+    case _ => false
+  }
+
+  override def subst() = TTuple(types.map(t => t.subst()))
+
+  override def _pretty(sb: StringBuilder, indent: Int, compact: Boolean) {
+    sb.append("Tuple[")
```

Comment:
I like parens here, how about you?

Reply:
I wanted to be consistent with the other types. We were using "(" for type parameters like Locus and GenomeReference. `Tuple[TInt32, TInt32]` is consistent with Scala tuples.

Reply:
fine with this.